mcs/class/System/Mono.UnityTls/UnityTlsContext.cs

```diff
@@ -0,0 +1,370 @@
+				Mono.Unity.Debug.CheckAndThrow (errorState, "Failed to set trace callback");
+			}
+
+			m_HasContext = true;
+		}
+
+		public override bool HasContext {
+			get { return m_HasContext; }
+		}
+
+		public override bool IsAuthenticated {
+			get { return m_IsAuthenticated; }
+		}
+
+		public override MonoTlsConnectionInfo ConnectionInfo {
+			get { return m_Connectioninfo; }
+		}
+		internal override bool IsRemoteCertificateAvailable {
+			get { return m_RemoteCertificate != null; }
+		}
+		internal override X509Certificate LocalClientCertificate {
+			get { return m_LocalClientCertificate; }
+		}
+		public override X509Certificate RemoteCertificate {
+			get { return m_RemoteCertificate; }
+		}
+		public override TlsProtocols NegotiatedProtocol {
+			get { return ConnectionInfo.ProtocolVersion; }
+		}
+
+		public override void Flush ()
+		{
+			// NO-OP
+		}
+
+		public override (int ret, bool wantMore) Read (byte[] buffer, int offset, int count)
+		{
+			bool wouldBlock = false;
+			int numBytesRead = 0;
+
+			var errorState = UnityTls.NativeInterface.unitytls_errorstate_create ();
+			fixed (byte* bufferPtr = buffer) {
+				numBytesRead = UnityTls.NativeInterface.unitytls_tlsctx_read (m_TlsContext, bufferPtr + offset, count, &errorState);
+			}
+
+			if (errorState.code == UnityTls.unitytls_error_code.UNITYTLS_USER_WOULD_BLOCK)
+				wouldBlock = true;
+			else
+				Mono.Unity.Debug.CheckAndThrow (errorState, "Failed to read data from TLS context");
+
+			return (numBytesRead, wouldBlock);
+		}
+
+		public override (int ret, bool wantMore) Write (byte[] buffer, int offset, int count)
+		{
+			bool wouldBlock = false;
+			int numBytesWritten = 0;
+
+			var errorState = UnityTls.NativeInterface.unitytls_errorstate_create ();
+			fixed (byte* bufferPtr = buffer) {
+				numBytesWritten = UnityTls.NativeInterface.unitytls_tlsctx_write (m_TlsContext, bufferPtr + offset, count, &errorState);
+			}
+
+			if (errorState.code == UnityTls.unitytls_error_code.UNITYTLS_USER_WOULD_BLOCK)
+				wouldBlock = true;
+			else
+				Mono.Unity.Debug.CheckAndThrow (errorState, "Failed to write data to TLS context");
+
+			return (numBytesWritten, wouldBlock);
+		}
+
+		public override void Shutdown ()
+		{
+			// Destroy native UnityTls objects
+			UnityTls.NativeInterface.unitytls_tlsctx_free (m_TlsContext);
+			m_TlsContext = null;
+
+			m_HasContext = false;
+		}
+
+		protected override void Dispose (bool disposing)
+		{
+			try {
+				if (disposing)
+				{
+					Shutdown();
+
+					// reset states
+					m_LocalClientCertificate = null;
+					m_RemoteCertificate = null;
+					m_Connectioninfo = null;
+					m_IsAuthenticated = false;
+					m_HasContext = false;
+				}
+
+				m_handle.Free();
+
+			} finally {
+				base.Dispose (disposing);
+			}
+		}
+
+		public override void StartHandshake ()
+		{
+			// TODO: Client->Server authentification is not supported by UnityTls as of writing
```

Comment:
should we fix this before pushing to master?

Reply:
Is this feature fully functional and tested in UnityTls by now. If so, clearly I should add the necessary wiring and another integration test on the unity side of this PR!

Reply:
it is

Reply:
Together with SNI. I wonder how SNI is supported in Mono..

Reply:
Done and tested.